File: android/app/src/main/java/com/francescooddo/buddygrammar/core/LocalWordCorrector.kt
package com.francescooddo.buddygrammar.core

import kotlin.math.abs
import kotlin.math.hypot

/**
 * Conservative on-device typo correction with cheaper costs for transposed
 * letters and adjacent QWERTY keys.
 */
object LocalWordCorrector {
    fun bestCorrection(typedWord: String, candidates: Iterable<String> = WordList.words): String? {
        val source = typedWord.lowercase()
        if (source.length < 2) return null
        val maximumLengthDelta = if (source.length >= 6) 2 else 1
        val seen = HashSet<String>()
        var best: RankedCandidate? = null
        var runnerUp: RankedCandidate? = null

        for (candidate in candidates) {
            if (candidate.isEmpty() || abs(candidate.length - source.length) > maximumLengthDelta) {
                continue
            }
            val normalizedCandidate = candidate.lowercase()
            if (!seen.add(normalizedCandidate)) continue
            if (normalizedCandidate == source) return null
            // Prefix extensions belong in the completion lane and must never
            // be offered or auto-applied as typo corrections.
            if (normalizedCandidate.startsWith(source)) continue

            val ranked = RankedCandidate(
                word = candidate,
                distance = normalizedDistance(source, normalizedCandidate),
            )
            if (ranked.distance > acceptanceThreshold(maxOf(source.length, candidate.length))) {
                continue
            }
            when {
                best == null || ranked < best -> {
                    runnerUp = best
                    best = ranked
                }
                runnerUp == null || ranked < runnerUp -> runnerUp = ranked
            }
        }

        val winner = best ?: return null
        if (
            runnerUp != null &&
            abs(winner.distance - runnerUp.distance) < MINIMUM_SCORE_MARGIN
        ) return null
        return matchingCapitalization(typedWord, winner.word)
    }

    private fun normalizedDistance(source: String, target: String): Double {
        if (source.isEmpty() || target.isEmpty()) return 1.0

        var twoRowsBack = DoubleArray(target.length + 1)
        var previousRow = DoubleArray(target.length + 1) { index -> index * INSERT_DELETE_COST }
        for (sourceIndex in 1..source.length) {
            val currentRow = DoubleArray(target.length + 1)
            currentRow[0] = sourceIndex * INSERT_DELETE_COST
            for (targetIndex in 1..target.length) {
                val substitution = previousRow[targetIndex - 1] +
                    substitutionCost(source[sourceIndex - 1], target[targetIndex - 1])
                val deletion = previousRow[targetIndex] + INSERT_DELETE_COST
                val insertion = currentRow[targetIndex - 1] + INSERT_DELETE_COST
                var best = minOf(substitution, deletion, insertion)

                if (
                    sourceIndex > 1 && targetIndex > 1 &&
                    source[sourceIndex - 1] == target[targetIndex - 2] &&
                    source[sourceIndex - 2] == target[targetIndex - 1]
                ) {
                    best = minOf(best, twoRowsBack[targetIndex - 2] + TRANSPOSE_COST)
                }
                currentRow[targetIndex] = best
            }
            twoRowsBack = previousRow
            previousRow = currentRow
        }
        return previousRow[target.length] / maxOf(source.length, target.length)
    }

    private data class RankedCandidate(
        val word: String,
        val distance: Double,
    ) : Comparable<RankedCandidate> {
        override fun compareTo(other: RankedCandidate): Int =
            compareValuesBy(this, other, RankedCandidate::distance, { it.word.length }, { it.word })
    }

    private fun substitutionCost(lhs: Char, rhs: Char): Double {
        if (lhs == rhs) return 0.0
        val left = keyPositions[lhs] ?: return 1.0
        val right = keyPositions[rhs] ?: return 1.0
        return if (hypot(left.first - right.first, left.second - right.second) <= 1.3) {
            NEAR_KEY_COST
        } else {
            1.0
        }
    }

    private fun acceptanceThreshold(length: Int): Double = when (length) {
        in 0..3 -> 0.20
        in 4..5 -> 0.34
        else -> 0.40
    }

    private fun matchingCapitalization(source: String, correction: String): String = when {
        source.any(Char::isLetter) && source.all { !it.isLetter() || it.isUpperCase() } ->
            correction.uppercase()
        source.firstOrNull()?.isUpperCase() == true ->
            correction.replaceFirstChar { it.uppercaseChar() }
        else -> correction
    }

    private val keyPositions: Map<Char, Pair<Double, Double>> = buildMap {
        val rows = listOf("qwertyuiop" to 0.0, "asdfghjkl" to 0.25, "zxcvbnm" to 0.75)
        rows.forEachIndexed { rowIndex, (letters, offset) ->
            letters.forEachIndexed { columnIndex, character ->
                put(character, (columnIndex + offset) to rowIndex.toDouble())
            }
        }
    }

    private const val INSERT_DELETE_COST = 0.9
    private const val NEAR_KEY_COST = 0.45
    private const val TRANSPOSE_COST = 0.25
    private const val MINIMUM_SCORE_MARGIN = 0.04
}

File: BuddyGrammarKit/Sources/BuddyGrammarKit/LocalWordCorrector.swift
import Foundation

public enum LocalWordCorrector {
    public static func bestCorrection(
        for typedWord: String,
        candidates: [String]
    ) -> String? {
        guard let best = rankedCorrections(
            for: typedWord,
            candidates: candidates,
            limit: 1
        ).first else { return nil }
        return matchingCapitalization(of: typedWord, appliedTo: best.word)
    }

    /// Candidates that pass the acceptance thresholds for `typedWord`, best
    /// first. `contextAdjustment` may return a negative bonus that influences
    /// ordering only; it can never rescue a candidate past its distance
    /// threshold, so context re-ranks without loosening correctness gates.
    public static func rankedCorrections(
        for typedWord: String,
        candidates: [String],
        limit: Int,
        contextAdjustment: ((String) -> Double)? = nil
    ) -> [(word: String, distance: Double)] {
        guard limit > 0 else { return [] }
        let source = typedWord.lowercased()
        guard source.count >= 2 else { return [] }
        let maximumLengthDelta = source.count >= 6 ? 2 : 1
        var seen = Set<String>()
        var accepted: [RankedCandidate] = []

        for (index, candidate) in candidates.enumerated() {
            guard !candidate.isEmpty,
                  abs(candidate.count - source.count) <= maximumLengthDelta else {
                continue
            }
            let normalized = candidate.lowercased()
            guard seen.insert(normalized).inserted else { continue }
            // An exact candidate means the typed word is already a known
            // spelling; no correction should be proposed at all.
            if normalized == source { return [] }
            let ranked = RankedCandidate(
                word: candidate,
                distance: normalizedDistance(from: source, to: normalized),
                index: index,
                adjustment: contextAdjustment?(normalized) ?? 0
            )
            guard ranked.distance <= acceptanceThreshold(
                forLength: max(source.count, candidate.count)
            ) else {
                continue
            }
            accepted.append(ranked)
        }

        let sorted = accepted.sorted { $0.isPreferred(over: $1) }
        return sorted.prefix(limit).map { ($0.word, $0.distance) }
    }

    private static func normalizedDistance(from source: String, to target: String) -> Double {
        weightedDistance(
            Array(source),
            Array(target)
        ) { substitutionCost($0, $1) }
    }

    /// Weighted Damerau-Levenshtein distance over lowercase ASCII letters,
    /// normalized by the longer length. QWERTY-adjacent substitutions and
    /// transpositions cost 0.45; other substitutions 1; insert/delete 0.9.
    /// Callers must guarantee both inputs are ASCII letters (see
    /// ``FuzzySpellingEngine/canonicalGeometry(for:)``).
    public static func weightedASCIIDistance(
        from source: [UInt8],
        to target: [UInt8]
    ) -> Double {
        var scratch = DistanceScratch(
            previousRow: [Double](repeating: 0, count: target.count + 1),
            currentRow: [Double](repeating: 0, count: target.count + 1),
            twoRowsBack: [Double](repeating: 0, count: target.count + 1)
        )
        return weightedASCIIDistance(from: source, to: target, scratch: &scratch)
    }

    /// Reusable DP buffers so batched lookups (one per keystroke pause) never
    /// allocate per candidate.
    public struct DistanceScratch {
        var previousRow: [Double]
        var currentRow: [Double]
        var twoRowsBack: [Double]

        init(previousRow: [Double], currentRow: [Double], twoRowsBack: [Double]) {
            self.previousRow = previousRow
            self.currentRow = currentRow
            self.twoRowsBack = twoRowsBack
        }
    }

    public static func weightedASCIIDistance(
        from source: [UInt8],
        to target: [UInt8],
        scratch: inout DistanceScratch
    ) -> Double {
        guard !source.isEmpty, !target.isEmpty else { return 1 }
        let requiredLength = target.count + 1
        let costs = flatSubstitutionCosts

        func ensure(_ row: inout [Double]) {
            if row.count < requiredLength {
                row = [Double](repeating: 0, count: requiredLength)
            }
        }
        ensure(&scratch.previousRow)
        ensure(&scratch.currentRow)
        ensure(&scratch.twoRowsBack)

        let width = requiredLength
        for column in 0..<width {
            scratch.previousRow[column] = Double(column) * 0.9
        }

        for sourceIndex in 1...source.count {
            scratch.currentRow[0] = Double(sourceIndex) * 0.9
            let sourceBase = Int(source[sourceIndex - 1] - 97) * 26
            let canTranspose = sourceIndex > 1
            for targetIndex in 1..<width {
                let substitution = scratch.previousRow[targetIndex - 1]
                    + costs[sourceBase + Int(target[targetIndex - 1] - 97)]
                let deletion = scratch.previousRow[targetIndex] + 0.9
                let insertion = scratch.currentRow[targetIndex - 1] + 0.9
                var best = substitution < deletion ? substitution : deletion
                if insertion < best { best = insertion }

                if canTranspose,
                   targetIndex > 1,
                   source[sourceIndex - 1] == target[targetIndex - 2],
                   source[sourceIndex - 2] == target[targetIndex - 1] {
                    let transposition = scratch.twoRowsBack[targetIndex - 2] + 0.45
                    if transposition < best { best = transposition }
                }
                scratch.currentRow[targetIndex] = best
            }
            // Rotate rows for the next pass.
            for column in 0..<width {
                scratch.twoRowsBack[column] = scratch.previousRow[column]
                scratch.previousRow[column] = scratch.currentRow[column]
            }
        }

        return scratch.previousRow[target.count]
            / Double(max(source.count, target.count))
    }

    private static func weightedDistance<Element: Equatable>(
        _ lhs: [Element],
        _ rhs: [Element],
        substitutionCost: (Element, Element) -> Double
    ) -> Double {
        guard !lhs.isEmpty, !rhs.isEmpty else { return 1 }

        var twoRowsBack = Array(repeating: 0.0, count: rhs.count + 1)
        var previousRow = (0...rhs.count).map { Double($0) * 0.9 }

        for lhsIndex in 1...lhs.count {
            var currentRow = Array(repeating: 0.0, count: rhs.count + 1)
            currentRow[0] = Double(lhsIndex) * 0.9
            for rhsIndex in 1...rhs.count {
                let substitution = previousRow[rhsIndex - 1]
                    + substitutionCost(lhs[lhsIndex - 1], rhs[rhsIndex - 1])
                let deletion = previousRow[rhsIndex] + 0.9
                let insertion = currentRow[rhsIndex - 1] + 0.9
                var best = min(substitution, deletion, insertion)

                if lhsIndex > 1,
                   rhsIndex > 1,
                   lhs[lhsIndex - 1] == rhs[rhsIndex - 2],
                   lhs[lhsIndex - 2] == rhs[rhsIndex - 1] {
                    best = min(best, twoRowsBack[rhsIndex - 2] + 0.45)
                }
                currentRow[rhsIndex] = best
            }
            twoRowsBack = previousRow
            previousRow = currentRow
        }

        return previousRow[rhs.count] / Double(max(lhs.count, rhs.count))
    }

    private struct RankedCandidate {
        let word: String
        let distance: Double
        let index: Int
        let adjustment: Double

        /// Ordering key: adjusted distance first (context bonus participates),
        /// then raw distance, then shorter word, then candidate order.
        func isPreferred(over other: Self) -> Bool {
            let lhsKey = distance + adjustment
            let rhsKey = other.distance + other.adjustment
            if abs(lhsKey - rhsKey) > 0.000_001 { return lhsKey < rhsKey }
            if abs(distance - other.distance) > 0.000_001 { return distance < other.distance }
            if word.count != other.word.count { return word.count < other.word.count }
            return index < other.index
        }
    }

    private static func substitutionCost(_ lhs: Character, _ rhs: Character) -> Double {
        guard lhs != rhs else { return 0 }
        guard let lhsScalar = lhs.asciiValue,
              let rhsScalar = rhs.asciiValue,
              areQwertyNeighbors(lhsScalar, rhsScalar) else {
            return 1
        }
        return 0.45
    }

    /// Precomputed QWERTY substitution costs for ASCII letters, indexed by
    /// `(lhs - 97) * 26 + (rhs - 97)`, so the DP inner loop stays branch-light.
    private static let flatSubstitutionCosts: [Double] = {
        var costs = [Double](repeating: 1, count: 26 * 26)
        for lhs in 0..<26 {
            for rhs in 0..<26 {
                let cost: Double
                if lhs == rhs {
                    cost = 0
                } else if areQwertyNeighbors(UInt8(lhs + 97), UInt8(rhs + 97)) {
                    cost = 0.45
                } else {
                    cost = 1
                }
                costs[lhs * 26 + rhs] = cost
            }
        }
        return costs
    }()

    private static func areQwertyNeighbors(_ lhs: UInt8, _ rhs: UInt8) -> Bool {
        guard let lhsPosition = asciiKeyPositions[lhs],
              let rhsPosition = asciiKeyPositions[rhs] else {
            return false
        }
        let horizontal = (Double(lhsPosition.column) + lhsPosition.rowOffset)
            - (Double(rhsPosition.column) + rhsPosition.rowOffset)
        let vertical = Double(lhsPosition.row) - Double(rhsPosition.row)
        return hypot(horizontal, vertical) <= 1.3
    }

    static func acceptanceThreshold(forLength length: Int) -> Double {
        switch length {
        case ...3: 0.2
        case 4...5: 0.34
        default: 0.4
        }
    }

    private static func matchingCapitalization(of source: String, appliedTo correction: String) -> String {
        if source == source.uppercased(), source.rangeOfCharacter(from: .letters) != nil {
            return correction.uppercased()
        }
        guard source.first?.isUppercase == true, let first = correction.first else {
            return correction
        }
        return first.uppercased() + correction.dropFirst()
    }

    private static let asciiKeyPositions: [UInt8: (row: Int, column: Int, rowOffset: Double)] = {
        let rows: [(String, Int, Double)] = [
            ("qwertyuiop", 0, 0),
            ("asdfghjkl", 1, 0.25),
            ("zxcvbnm", 2, 0.75),
        ]
        var positions: [UInt8: (row: Int, column: Int, rowOffset: Double)] = [:]
        for (row, rowIndex, rowOffset) in rows {
            for (columnIndex, character) in row.utf8.enumerated() {
                positions[character] = (rowIndex, columnIndex, rowOffset)
            }
        }
        return positions
    }()
}
